### PR TITLE
Reference date

### PR DESCRIPTION
**Pull request by garcipat**
The SystemClock is static and this is problematic if you have different reference dates. I wanted to provide the option to set the reference date on a faker as with the localization. The default is like before, nothing changes, therefore there should be no breaking change.

Example:
```
var f = new Faker();
f.ReferenceDate = new DateTime(2023, 12, 30, 12, 30, 0);
```

### CODE REVIEW

File: Source/Bogus/DataSets/Date.cs [hunk 1]
```diff
@@ -25,22 +27,26 @@ public partial class Date : DataSet
       /// <summary>
       /// Create a Date dataset
       /// </summary>
-      public Date(string locale = "en") : base(locale)
+      public Date(string locale = "en", Func<DateTime?> referenceDate = null) : base(locale)
```
Comment (304NotModified) [inline Source/Bogus/DataSets/Date.cs:30]: note: this is a binary breaking change

**Comment by 304NotModified** [comment]
Yes! Great work!

File: Source/Bogus/DataSets/Date.cs (hunk 1 above)
Comment (garcipat) [reply Source/Bogus/DataSets/Date.cs:30]: I think I had it that way previously and changed it. Let me revise why. It has been a moment.

File: Source/Bogus.Tests/LocaleTests/PtLocale.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public void name_tests()
       [Fact]
       public void date_tests()
       {
-         var d = new Date("pt_PT");
+         var d = new Date(locale: "pt_PT");
```
Comment (bchavez) [inline Source/Bogus.Tests/LocaleTests/PtLocale.cs:41]: also, you'll need to revert this change if it's not directly related to fixing the problem at hand.

superficial changes only create noise in the PR review/merge (and in the unlikely event, revert) process. PR change sets into this repo should only include the minimal amount of changes required to fix or correct a single problem.

coding style improvements would need to go under a different PR.

Comment (garcipat) [reply Source/Bogus.Tests/LocaleTests/PtLocale.cs:41]: I did only do the necessary changes. Maybe thats unintentional and a leftover from the pwssing a function over a second ctor.  I will chech it. thanks for the hint.

File: Source/Bogus/DataSets/Date.cs (hunk 1 above)
Comment (garcipat) [reply Source/Bogus/DataSets/Date.cs:30]: This does not work like that since both ctors only have optional arguments. I can move the referencedate as a non optional argument to ste start on the second ctor. I dont like that so much but i guess its better.

Comment (garcipat) [reply Source/Bogus/DataSets/Date.cs]: Added the overload

File: Source/Bogus.Tests/LocaleTests/PtLocale.cs (hunk 2 above)
Comment (garcipat) [reply Source/Bogus.Tests/LocaleTests/PtLocale.cs]: Is reverted.

**Comment by garcipat** [comment]
@bchavez I also don't know if on GitHub people resolve the covnersations after an issue has been solved or I wait untilf the person mentioning it is resolves it himself. I started contributing just recently.

**Comment by 304NotModified** [comment]
> @bchavez I also don't know if on GitHub people resolve the covnersations after an issue has been solved or I wait untilf the person mentioning it is resolves it himself. I started contributing just recently.

If it's clear that is has been fixed, it's imo preferred that you press resolve:)

**Comment by garcipat** [comment]
> > @bchavez I also don't know if on GitHub people resolve the covnersations after an issue has been solved or I wait untilf the person mentioning it is resolves it himself. I started contributing just recently.
> 
> If it's clear that is has been fixed, it's imo preferred that you press resolve:)

Thanks for the clarification :)

**Review by bchavez — CHANGES_REQUESTED** [review]
Hello there; hope your days are going well.

Upon further review, I think there's a bug with this PR because the changes missed the `Date.net60.cs` .NET 6 specific `TimeOnly`/`DateOnly` methods which still read from `SystemClock()` that should have read from `GetDate()`.

Alternatively, I put together PR #500 which fixes the problem and also plumbs the local time ref to `Faker<T>` and other places where it's probably important API / consistency / ergonomic-wise.

Please post some thoughts on #500.

And for this PR I'm marking this PR as a review block because of the missed `SystemClock()` bugs in the `Date.net60.cs` file and until we discuss more.

Let me know your thoughts. Thanks.

**Comment by garcipat** [comment]
> Hello there; hope your days are going well.
> 
> Upon further review, I think there's a bug with this PR because the changes missed the `Date.net60.cs` .NET 6 specific `TimeOnly`/`DateOnly` methods which still read from `SystemClock()` that should have read from `GetDate()`.
> 
> Alternatively, I put together PR #500 which fixes the problem and also plumbs the local time ref to `Faker<T>` and other places where it's probably important API / consistency / ergonomic-wise.
> 
> Please post some thoughts on #500.
> 
> And for this PR I'm marking this PR as a review block because of the missed `SystemClock()` bugs in the `Date.net60.cs` file and until we discuss more.
> 
> Let me know your thoughts. Thanks.

Good catch. I will check where the SystemClock is used in egneral and observe if there is any other place I maybe missed.

**Comment by 304NotModified** [comment]
@garcipat @bchavez 
Thanks for all the work!  

I'm now a bit confused. What is the latest PR of this feature? This one or #500? Or aren't those the same?

**Comment by garcipat** [comment]
> @garcipat @bchavez Thanks for all the work!
> 
> I'm now a bit confused. What is the latest PR of this feature? This one or #500? Or aren't those the same?

@304NotModified He extended on my idea and created his own branch in the main repo. I now don't know what to do with this one. I can't help or do anything now 😅

**Comment by bchavez** [comment]
@garcipat @304NotModified . No worries, I'll make a decision as soon as I review the most recent changes in this 492 PR.

**Comment by garcipat** [comment]
> @garcipat @304NotModified . No worries, I'll make a decision as soon as I review the most recent changes in this 492 PR.

Alright. Thank you.

**Comment by bchavez** [comment]
Alright, I think we'll go with PR #500 and put our energies there now that things are starting to take shape.

**Comment by garcipat** [comment]
> Alright, I think we'll go with PR #500 and put our energies there now that things are starting to take shape.

But I will then take over to #500 what you already did there. Naming and such.
